kernel/contracts/ZepCore.sol

```diff
@@ -0,0 +1,85 @@
+pragma solidity ^0.4.18;
+
+import "./ZepToken.sol";
+import "./KernelInstance.sol";
+import "./KernelRegistry.sol";
+import "zeppelin-solidity/contracts/math/SafeMath.sol";
+
+contract ZepCore {
+  using SafeMath for uint256;
+
+  event Staked(address indexed user, uint256 amount, uint256 total, bytes data);
+  event Unstaked(address indexed user, uint256 amount, uint256 total, bytes data);
+
+  uint256 public newVersionCost;
+  uint256 public developerFraction;
+
+  ZepToken private _token;
+  KernelRegistry private _registry;
+  uint256 private _totalStaked;
+  mapping(address => uint256) private _instanceVouches;
+  mapping(address => mapping (address => uint256)) private _userVouches;
+
+  function ZepCore(uint256 _newVersionCost, uint256 _developerFraction) public {
+    _token = new ZepToken();
+    _registry = new KernelRegistry();
+    developerFraction = _developerFraction;
+    newVersionCost = _newVersionCost;
+    // TODO: we need to think how we are going to manage variable costs to propose new versions
+  }
+
+  function register(KernelInstance instance) public {
+    registry().addInstance(instance);
+    token().transferFrom(msg.sender, this, newVersionCost);
+    token().burn(newVersionCost);
+    // TODO: should we build a burnFrom or sth like that?
+  }
+
+  function transferStake(KernelInstance from, KernelInstance to, uint256 amount, bytes data) public {
+    // TODO: an unstake followed by a stake call won't work
+    require(false);
+  }
+
+  function stake(KernelInstance instance, uint256 amount, bytes data) public {
+    token().transferFrom(msg.sender, this, amount);
+    uint256 developerPayout = amount.div(developerFraction);
+    require(developerPayout > 0);
+    token().transfer(instance.developer(), developerPayout);
+    // TODO: check how we can manage remainders in a better way
+
+    uint256 stakedAmount = amount.sub(developerPayout);
+    _totalStaked = totalStaked().add(stakedAmount);
+    _instanceVouches[instance] = totalStakedFor(instance).add(stakedAmount);
+    _userVouches[msg.sender][instance] = _userVouches[msg.sender][instance].add(stakedAmount);
+
+    Staked(msg.sender, stakedAmount, totalStakedFor(instance), data);
```

Comment:
The event never reports which instance was staked for, which is critical here. Maybe we should diverge from ERC900 in the events signature?

Reply:
Added, we've diverged from ERC900 quite a bit anyway, precisely because it doesn't emphasize the difference between the staker and the object of the stake.